python_modules/dagster/dagster/_core/external_execution/task.py

```diff
@@ -167,19 +168,27 @@ def _output_context_manager(self) -> ContextManager[OutputIOYield]:
     @contextmanager
     def _stdio_output(self) -> Iterator[OutputIOYield]:
         read_fd, write_fd = os.pipe()
-        thread = self._default_output_thread(read_fd)
-        yield write_fd, {}
-        os.close(write_fd)
-        thread.join()
+        is_task_complete = Event()
+        thread = self._default_output_thread(read_fd, is_task_complete)
+        try:
+            yield write_fd, {}
+        finally:
+            os.close(write_fd)
+            is_task_complete.set()
+            thread.join()
 
     @contextmanager
     def _file_output(self) -> Iterator[OutputIOYield]:
         path = self._prepare_io_path(self._output_path, "output")
         env = {DAGSTER_EXTERNAL_ENV_KEYS["output"]: path}
+        open(path, "w").close()  # create file
+        is_task_complete = Event()
+        thread = self._default_output_thread(path, is_task_complete)
```

Comment:
It causes type errors in `finally` because any variable defined within `try` is assumed to be possibly unbound in `finally`. And there is no need to clean up the `is_task_complete` event in a `finally`, and the `thread` won't exist to be cleaned up if the second line fails.